assembly/src/assembler/mod.rs

```diff
@@ -22,6 +23,98 @@ mod tests;
 
 type ProcedureCache = BTreeMap<ProcedureId, Procedure>;
 
+// MODULE PROVIDER
+// ================================================================================================
+
+/// A module provider that will store the procedure ASTs used during the compilation process.
+#[derive(Debug, Default, Clone, PartialEq, Eq)]
+pub struct ModuleProvider {
+    modules: Vec<Module>,
+    /// Map from procedure id to module index + proc index in the module.
+    procedures: BTreeMap<ProcedureId, (usize, usize)>,
+}
+
+impl ModuleProvider {
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Fetch a module that contains the provided procedure id.
+    pub fn get_module(&self, id: &ProcedureId) -> Option<&Module> {
+        // this will panic only if there is a bug in `Self::add_module`.
+        self.procedures.get(id).map(|(i, _)| &self.modules[*i])
+    }
+
+    /// Fetch a module AST by its absolute path.
+    ///
+    /// This function might be useful in future refactoring.
+    #[allow(dead_code)]
+    #[deprecated(since = "0.4.0", note = "module provider should have a single mapping")]
+    pub fn get_module_by_path(&self, path: &AbsolutePath) -> Option<&ModuleAst> {
+        let module_path = path;
+        self.modules
+            .iter()
+            .find_map(|Module { path, ast }| (module_path == path).then_some(ast))
+    }
+
+    /// Fetch a procedure AST by its id.
+    ///
+    /// This function might be useful in future refactoring.
+    #[allow(dead_code)]
+    #[deprecated(since = "0.4.0", note = "module provider should have a single mapping")]
+    pub fn get_procedure(&self, id: &ProcedureId) -> Option<&ProcedureAst> {
+        self.procedures
+            .get(id)
+            // this will panic only if there is a bug in `Self::add_module`.
+            .map(|(module, proc)| &self.modules[*module].ast.local_procs[*proc])
+    }
+
+    /// Fetch a procedure local procedure index by its id.
+    ///
+    /// This function might be useful in future refactoring.
+    #[allow(dead_code)]
+    #[deprecated(since = "0.4.0", note = "module provider should have a single mapping")]
+    pub fn get_procedure_index(&self, id: &ProcedureId) -> Option<usize> {
+        self.procedures.get(id).map(|(_, proc)| *proc)
+    }
```

Comment:
I don't think we need to keep this code around since it was only used internally anyway. Let's delete it.

Reply:
Indeed, I had doubts about that. Just kept to leave the procedures map more flexible (with a pointer both to the module ast and its procedure). But if we need it in the future, we can trivially add it